### PR TITLE
Move ini_set calls before a new session is started

### LINKED ISSUE #31524: PHP 7.2.5 ini_set(): A session is active
Everything seems working fine but if I check the owncloud log I see a lot of this errors:
"message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}

I recently upgraded PHP from version 5.5 to 7.2.5.
I already checked openssl module, it is installed and configured.

### Steps to reproduce
1. check the owncloud log file
2.
3.

### Expected behaviour
Tell us what should happen

### Actual behaviour
Tell us what happens instead

### Server configuration
**Operating system**:
Opensuse 42.3

**Web server:**
Apache 2.4.23

**Database:**
MariaDB

**PHP version:**
7.2.5

**ownCloud version:** (see ownCloud admin page)
10.0.8.5
**Updated from an older ownCloud or fresh install:**
Updated
**Where did you install ownCloud from:**
Manual
**Signing status (ownCloud 9.0 and above):**

```
Login as admin user into your ownCloud and access 
http://example.com/index.php/settings/integrity/failed 
paste the results into https://gist.github.com/ and puth the link here.

No errors have been found.
```

**The content of config/config.php:**

```
Log in to the web-UI with an administrator account and click on
'admin' -> 'Generate Config Report' -> 'Download ownCloud config report'
This report includes the config.php settings, the list of activated apps
and other details in a well sanitized form.

or 

If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your ownCloud installation folder

*ATTENTION:* Do not post your config.php file in public as is. Please use one of the above
methods whenever possible. Both, the generated reports from the web-ui and from occ config:list
consistently remove sensitive data. You still may want to review the report before sending.
If done manually then it is critical for your own privacy to dilligently
remove *all* host names, passwords, usernames, salts and other credentials before posting.
You should assume that attackers find such information and will use them against your systems.
```

**List of activated apps:**

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your ownCloud installation folder.
```

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** yes/no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

#### LDAP configuration (delete this part if not used)

```
With access to your command line run e.g.:
sudo -u www-data php occ ldap:show-config
from within your ownCloud installation folder

Without access to your command line download the data/owncloud.db to your local
computer or access your SQL server remotely and run the select query:
SELECT * FROM `oc_appconfig` WHERE `appid` = 'user_ldap';


Eventually replace sensitive data as the name/IP-address of your LDAP server or groups.
```

### Client configuration
**Browser:**

**Operating system:**

#### Web server error log
```
Insert your webserver log here
```

#### ownCloud log (data/owncloud.log)
```
Insert your ownCloud log here
```
{"reqId":"700f019b-b6db-4c93-8f03-0c2f6ad62563","level":3,"time":"2018-05-24T12:24:18+00:00","remoteAddr":"","user":"--","app":"PHP","method":"PROPFIND","url":"\/owncloud\/remote.php\/dav\/files\/marma\/","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"q3fAF87ySupiF9aQaY29","level":3,"time":"2018-05-24T12:24:44+00:00","remoteAddr":"","user":"--","app":"PHP","method":"GET","url":"\/owncloud\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"78f27b57-ddd1-4f30-b695-b0d5259b2d69","level":3,"time":"2018-05-24T12:24:48+00:00","remoteAddr":"","user":"--","app":"PHP","method":"PROPFIND","url":"\/owncloud\/remote.php\/dav\/files\/marma\/","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"U6Cs1ADV8C0jmS4rB6gy","level":3,"time":"2018-05-24T12:24:59+00:00","remoteAddr":"","user":"--","app":"PHP","method":"GET","url":"\/owncloud\/index.php\/settings\/integrity\/failed","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"88c28a7b-f12e-4477-9ff2-92aa6ae91f81","level":3,"time":"2018-05-24T12:25:16+00:00","remoteAddr":"","user":"--","app":"PHP","method":"GET","url":"\/owncloud\/ocs\/v2.php\/apps\/notifications\/api\/v1\/notifications?format=json","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"1bbe77be-3c3c-4131-9d45-5364b910a10b","level":3,"time":"2018-05-24T12:25:18+00:00","remoteAddr":"","user":"--","app":"PHP","method":"PROPFIND","url":"\/owncloud\/remote.php\/dav\/files\/marma\/","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}
{"reqId":"f3ce1561-0936-484f-8adc-22f85c1937ea","level":3,"time":"2018-05-24T12:25:48+00:00","remoteAddr":"","user":"--","app":"PHP","method":"PROPFIND","url":"\/owncloud\/remote.php\/dav\/files\/marma\/","message":"ini_set(): A session is active. You cannot change the session module's ini settings at this time at \/srv\/www\/htdocs\/owncloud\/lib\/base.php#596"}

#### Browser log
```
Insert your browser log here, this could for example include:

a) The javascript console log
b) The network log 
c) ...
```

### PR DESCRIPTION
**Pull request by DeepDiver1975**
## Description
Alternative approach to https://github.com/owncloud/core/pull/31525

Inspired by https://github.com/joomla/joomla-cms/pull/15742/files

## Related Issue
- Fixes #31524

## How Has This Been Tested?
- manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

**Comment by patrickjahns** [comment]
@DeepDiver1975 

I don't quite understand why you move non-session related ini_set methods also into the session part

Wondering about the `Request` object "helper" -> we could follow the symfony approach where we create a request object early on before the "Kernel" is bootstrapped

https://symfony.com/doc/current/components/http_foundation.html#request

Example: https://github.com/Sylius/Sylius/blob/master/web/app.php#L23-L27

**Comment by DeepDiver1975** [comment]
@patrickjahns please review once more - thx

**Comment by patrickjahns** [comment]
> Our request object is already too big and has too much dependencies - initializing the request object at the wrong point in time leads to severe issues .

Hence the idea to have a slim `Request::createFromGlobals();` -> all other things on the request should be set over time

If we think about a Middleware Handler approach, with each Middleware the request object could be either extended or wrapped to be passend down the line with more information.

So we get rid of the request object dependencies and have a clear request -> response cycle

File: lib/private/Session/Internal.php [hunk 1]
```diff
@@ -154,4 +156,40 @@ private function validateSession() {
 			throw new SessionNotAvailableException('Session has been closed - no further changes to the session are allowed');
 		}
 	}
+
+	private function start(): void {
+		if (@\session_id() === '') {
+			// prevents javascript from accessing php session cookies
```
Comment (patrickjahns) [inline lib/private/Session/Internal.php]: I know this is just a code move - but do we have more information why this is? Wondering if this still holds true today

**Comment by DeepDiver1975** [comment]
> Hence the idea to have a slim Request::createFromGlobals(); -> all other things on the request should be set over time
> 
> If we think about a Middleware Handler approach, with each Middleware the request object could be either extended or wrapped to be passend down the line with more information.
> 
> So we get rid of the request object dependencies and have a clear request -> response cycle

fully understood the concept - no need to explain - question is how to do the transition. We basically need to remove the csrf check from IRequest - then the dependency cylce issue is gone.

Not that critical ... we need to double check -> different PR

File: lib/private/Session/Internal.php (hunk 1 above)
Comment (DeepDiver1975) [reply lib/private/Session/Internal.php]: This comment is basically just copied over from the php docs - http://php.net/manual/en/session.security.ini.php

**Comment by patrickjahns** [comment]
> Not that critical ... we need to double check -> different PR 

👍 

So for this PR to merge - I wonder if can get ui tests to run with https. That way the cookie change to move to https would be tested in the sessions. 


Needs some investigation @phil-davis @individual-it

**Comment by individual-it** [comment]
@patrickjahns there are ways to auto accept self signed SSL certificates in selenium, would need a practical try to see if they work for us

**Comment by individual-it** [comment]
opened issue for running tests with SSL: https://github.com/owncloud/QA/issues/581

**Comment by patrickjahns** [comment]
@DeepDiver1975 
For approving I really would like to see https running - otherwise the change in behavior wouldn't be tested

**Review by miqrogroove — APPROVED** [review]
Works for me.

**Comment by PVince81** [comment]
@patrickjahns merge for master but block backport until acceptance tests are there ?

having it in master will at least have this run on developers envs which might confirm further that it works

**Review by patrickjahns — APPROVED** [review]
Merging for master 👍

**Comment by patrickjahns** [comment]
@PVince81 - lets go ahead

**Comment by PVince81** [comment]
stable10: https://github.com/owncloud/core/pull/32538

**Pull request merged**